### PR TITLE
Sims 13. Add Mood Status to living mobs.

### PR DESCRIPTION
**Pull request by Xackii**
## What Does This PR Do
 Add Mood status to living mobs. mood status is divided into mood gained from eating food, mood dependent on the beauty of the room, and unique mood and screws that can be obtained from different aspects of the station. Eating food of the highest quality increases mood, as does eating food that appeals to the player's species. Blood, organs and other debris reduce the beauty of the room, which lowers the player's mood when he enters this room, while beautiful objects such as flower pots, posters, animals, etc. raise. There are also some unique events for species (there are not many of them, more should be added in the future).

## Why It's Good For The Game
 More RP and immersion in the atmosphere of the station and roleplaying the character. Gameplay mechanics aimed at keeping the room clean and eating better food. Sims 13 RP. Mood status can also be disabled in the config.

## How does mood affect the player?
 A good mood increases the player's movement speed(0.1/0.25/0.5). A bad mood reduces his maximum HP(25,50).

## Also Video that showcase how it work
 https://www.youtube.com/watch?v=eM9JaNz9w8c

## Changelog
:cl:
add: Added Mood Status
/:cl:

### CODE REVIEW

**Comment by PhantornRU** [comment]
It sounds good. I like the implementation of events through datums, this allows you to flexibly configure different sources that change moods.
But the areas on which the mood also depends? Hm. Acceptable

**Comment by Xackii** [comment]
> It sounds good. I like the implementation of events through datums, this allows you to flexibly configure different sources that change moods. But the areas on which the mood also depends? Hm. Acceptable

all possible events are recorded in the mood events folder and divided into categories. for example, events in the category generic negative events are assigned via add_mud_event in the file where the event itself occurs. For example, the mud event slipped is recorded in the slippery element when the mob slips.

**Comment by EmeraldCandy** [comment]
_**opinion post**_
I dont think making mood mechanical would help with RP. people who rp will know how their character feels, and those who dont rp will almost certainly just ignore it. Especially since its mechanical, rp is unlikely to change it, thus making the mood mostly tied to arbitrary side-tasks. Ive not been a fan of this system even on heavier roleplay servers, I dont see people actually use the device since there is a rift between a characters actual feelings and what an rp tool says they should feel. As an example, you could just watch someone get murdered in front of you, or get told your best friend was just taken by xenos, but you could be happy because you ate some cake in a clean room. I dont think this kind of feature fits well in paradise's environment

**Comment by DGamerL** [comment]
Heya, thanks for making the time to open this PR!
A few pointers though:
- The PR template exists for a reason, please add the sections where you show you've tested it, and some images or videos that showcase the system
- For large and involved PRs, it's best to talk to our design team beforehand so they have time to discuss and give suggestions before opening the PR, and to avoid wasted time!

File: code/__DEFINES/mob_defines.dm [hunk 1]
```diff
@@ -93,6 +93,90 @@
...
+//Moods levels for humans
+#define MOOD_LEVEL_HAPPY4 9
+#define MOOD_LEVEL_HAPPY3 8
+#define MOOD_LEVEL_HAPPY2 7
+#define MOOD_LEVEL_HAPPY1 6
+#define MOOD_LEVEL_NEUTRAL 5
+#define MOOD_LEVEL_SAD1 4
+#define MOOD_LEVEL_SAD2 3
+#define MOOD_LEVEL_SAD3 2
+#define MOOD_LEVEL_SAD4 1
```
Comment (DGamerL) [inline code/__DEFINES/mob_defines.dm:123]: Please make the comments on these mood defines more clear what they are used for

**Comment by CinnamonSnowball** [comment]
This PR effectively stealthily raises the base movement speed of humans (as having high mood would likely become the "standard"), due to these changes *completely unmentioned in the PR description*:

At highest mood, this is effectively Nuka Cola/Umbrae Cloak of Darkness speed boost.

**Comment by GDNgit** [comment]
I think moodlets are an interesting concept but I see some design issues with this pr.

You really should more explicitly state what major changes are done in this pr. Changing speed dependent on mood is an incredibly impactful change and not putting it in the PR description is questionable as hell. More than any other codebase speed matters the most here and historically speed altering mechanics have had poor impacts on gameplay (see the slowdown pull PR, old meth or old adrenals).

Just looking over the code suggests that this mechanic is extremely easy to break, stuff like walking into a room with a pile of one decal in a corner and having your speed drop by a lot. Plants also giving 700 beauty is questionable considering deltastation is a map people play on. This is going to be difficult to balance and I suspect it to be extremely cheesable.

If this was just an optional thing with no gameplay impact it could be a bit better maybe? At best this mechanic seems to potentially be very funny as you watch your mood plummet after glancing at blood. A new meta has formed and it's averting your eyes from all sin

(PS: Don't be afraid to reach out for help on this. Some people will actually just end up just being dicks about it but I think this can work. Either way it seems like wayy too much for a solo project.)

**Comment by Xackii** [comment]
> I think moodlets are an interesting concept but I see some design issues with this pr.
> 
> You really should more explicitly state what major changes are done in this pr. Changing speed dependent on mood is an incredibly impactful change and not putting it in the PR description is questionable as hell. More than any other codebase speed matters the most here and historically speed altering mechanics have had poor impacts on gameplay (see the slowdown pull PR, old meth or old adrenals).
> 
> Just looking over the code suggests that this mechanic is extremely easy to break, stuff like walking into a room with a pile of one decal in a corner and having your speed drop by a lot. Plants also giving 700 beauty is questionable considering deltastation is a map people play on. This is going to be difficult to balance and I suspect it to be extremely cheesable.
> 
> If this was just an optional thing with no gameplay impact it could be a bit better maybe? At best this mechanic seems to potentially be very funny as you watch your mood plummet after glancing at blood. A new meta has formed and it's averting your eyes from all sin
> 
> (PS: Don't be afraid to reach out for help on this. Some people will actually just end up just being dicks about it but I think this can work. Either way it seems like wayy too much for a solo project.)



> I think moodlets are an interesting concept but I see some design issues with this pr.
> 
> You really should more explicitly state what major changes are done in this pr. Changing speed dependent on mood is an incredibly impactful change and not putting it in the PR description is questionable as hell. More than any other codebase speed matters the most here and historically speed altering mechanics have had poor impacts on gameplay (see the slowdown pull PR, old meth or old adrenals).
> 
> Just looking over the code suggests that this mechanic is extremely easy to break, stuff like walking into a room with a pile of one decal in a corner and having your speed drop by a lot. Plants also giving 700 beauty is questionable considering deltastation is a map people play on. This is going to be difficult to balance and I suspect it to be extremely cheesable.
> 
> If this was just an optional thing with no gameplay impact it could be a bit better maybe? At best this mechanic seems to potentially be very funny as you watch your mood plummet after glancing at blood. A new meta has formed and it's averting your eyes from all sin
> 
> (PS: Don't be afraid to reach out for help on this. Some people will actually just end up just being dicks about it but I think this can work. Either way it seems like wayy too much for a solo project.)

Regarding the abuse of the beauty of the room: in muda two parameters are considered which can be called as: 1 (mood). then what level of comfort the character is experiencing now and 2 (sanity). this is what level of comfort the character actually has. That is, if you were in a good mood and you entered a room full of blood and guts, then your first parameter will drop to a sad muda, but the parameter that is responsible for issuing buffs and debuffs is 2 and this parameter will descend over time (by tick) to level 1. So if you run past a dirty room, it won’t really affect your sanity (probably you should have written this in the pre :) ). It was also worth describing how good and bad mood affects the character. I should also have written about how good and bad mood affects the character (I completely forgot :) ). as already said above, a good mood increases the speed of movement (if such an idea is not appreciated, then you can either reduce the buffs to speed or come up with something more interesting) and a bad mood reduces the maximum amount of health of the character (something like the will to live or resistance to pain). If there is still a need, I can add these additions to the main pr

File: code/__DEFINES/mob_defines.dm (hunk 1 above)
Comment (Xackii) [reply code/__DEFINES/mob_defines.dm:123]: I'll do it

**Review by Contrabang — CHANGES_REQUESTED** [review]
Changelog, testing section, and PR description all need to be expanded to describe more accurately what the PR does (it is not expected for every contributor to come here and have to read all the code to understand what is going on). Along with that, a PR of this size and change should have video(s) and screenshots attached.

Some priority things I think need mentioning in the PR:
- Mood affects character speed
- All races now have preferred food types, which will increase or decrease their mood.
- Sanity affects your max health and current health

**Comment by Mitchs98** [comment]
Very much against this for reasons stated above. Mechanics shouldn't exist that railroad you into RPing your character a certain way; diet, what your character "should" be reacting to and how they "should" be reacting, even what your character "should" find appealing in terms of decor and environment. Either follow all of this or suffer debuffs. 

This won't improve RP at all, this will force people to do certain things or act certain ways or look at certain things or suffer debuffs. If anything, forcing people to react a certain way will _negatively impact_ roleplay by making it more generic/strongarming people into acting certain ways. Players should be free to play their character how they wish within reason. Is part of it LRP? Sure. Are we a HRP server? No. We're MRP. A bit of LRP or not serious roleplay is just fine. 

Trying to _force_ people into acting a certain way that truly isn't malicious just because it doesn't fit your definition of what should be going on isn't a way to go about it. You say people can ignore moodlets...but you can't? Not if it has a mechanical impact.

I am _firmly_ of the opinion that if someone's actions aren't malicious or somehow directly negatively impacting someone's experience or just outright batshit insane then there's very little reason to try and do something like this. Do I have issues and opinions against what some people do? Sure, and I'll mention it and all that nice stuff. But most of the time it's like, my opinion man. Some people enjoy it, I might not, some others might not. At the end of the day what's good roleplay is subjective whether I or anyone else likes it or not, regardless of what I or anyone else says is acceptable or not. Barring staff members, of course.

**Comment by Carthusia** [comment]
Never been a fan of the mood systems in SS13

Theres nothing worse than the disconnect of you talking with people, being fairly happy and chipper, but then in the corner it says "YOU ARE DEPRESSED" or "YOU ARE FEELING SAD!" when your clearly not

File: code/datums/elements/beauty.dm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		if(current_area && !current_area.outdoors)
+			current_area.totalbeauty -= beauty
+			current_area.update_beauty()
+		if(!beauty_counter[source])
+			. = ..()
+			UnregisterSignal(source, list(COMSIG_ENTER_AREA, COMSIG_EXIT_AREA))
+			beauty_counter -= source
+			/*var/atom/movable/movable_source = source
+			if(istype(movable_source))
+				movable_source.lose_area_sensitivity(BEAUTY_ELEMENT_TRAIT)*/
```
Comment (DGamerL) [inline code/datums/elements/beauty.dm:76]: Why is all this commented out stuff here?

File: code/datums/mood.dm [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+		if(MOOD_LEVEL_HAPPY4)
+			set_sanity(sanity + 0.6 * seconds_per_tick, SANITY_NEUTRAL, SANITY_MAXIMUM)
+	handle_nutrition()
+
+/datum/mood/proc/handle_mob_death(datum/source)
+	SIGNAL_HANDLER
+
+	if (last_stat == DEAD && mob_parent.stat != DEAD)
+		START_PROCESSING(SSmood, src)
+	else if (last_stat != DEAD && mob_parent.stat == DEAD)
```
Comment (DGamerL) [inline code/datums/mood.dm:100]:
```suggestion
	if(last_stat == DEAD && mob_parent.stat != DEAD)
		START_PROCESSING(SSmood, src)
	else if(last_stat != DEAD && mob_parent.stat == DEAD)
```

File: code/datums/mood.dm [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+/datum/mood/proc/handle_nutrition()
+	if (HAS_TRAIT(mob_parent, TRAIT_NOHUNGER))
+		clear_mood_event(MOOD_CATEGORY_NUTRITION)  // if you happen to switch species while hungry youre no longer hungy
+		return FALSE // no moods for nutrition
+	switch(mob_parent.nutrition)
+		if(NUTRITION_LEVEL_FULL to INFINITY)
+			add_mood_event(MOOD_CATEGORY_NUTRITION, /datum/mood_event/fat)
+		if(NUTRITION_LEVEL_WELL_FED to NUTRITION_LEVEL_FULL)
+			add_mood_event(MOOD_CATEGORY_NUTRITION, /datum/mood_event/wellfed)
+		if( NUTRITION_LEVEL_FED to NUTRITION_LEVEL_WELL_FED)
```
Comment (DGamerL) [inline code/datums/mood.dm:114]:
```suggestion
		if(NUTRITION_LEVEL_FED to NUTRITION_LEVEL_WELL_FED)
```

File: code/datums/mood.dm [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+		return
+
+	the_event.category = category
+	if (!isnull(mood_to_copy_from))
+		the_event.timeout = mood_to_copy_from.timeout
+	qdel(mood_to_copy_from)
+	mood_events[category] = the_event
+	update_mood()
+
+	if (the_event.timeout)
```
Comment (DGamerL) [inline code/datums/mood.dm:170]:
```suggestion
	if(!isnull(mood_to_copy_from))
		the_event.timeout = mood_to_copy_from.timeout
	qdel(mood_to_copy_from)
	mood_events[category] = the_event
	update_mood()

	if(the_event.timeout)
```

File: code/datums/mood.dm [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+		addtimer(CALLBACK(src, PROC_REF(clear_mood_event), category), the_event.timeout, (TIMER_UNIQUE|TIMER_OVERRIDE))
+
+/**
+ * Removes a mood event from the mob
+ *
+ * Arguments:
+ * * category - (Text) Removes the mood event with the given category
+ */
+/datum/mood/proc/clear_mood_event(category)
+	if (!istext(category))
```
Comment (DGamerL) [inline code/datums/mood.dm:180]:
```suggestion
	if(!istext(category))
```

**Comment by Fluffyzack21** [comment]
**Okay my opinion on this pr**

That is going to get abused the shit out of it, antags included, since they always want to be super happy, to get max heath and a speed buff, to be one step ahead of security and meanwhile a officer with debuffs will never get to them, I was okay with it if you know how to be always happy. But this will never be fucking used, people rp how they think there character is feeling instead of mood, so if you want my opinion on this PR, This PR will get the most hate from the community as a whole, but paradise is a mrp and little bit of lrp this is not TG

**Comment by farie82** [comment]
Please add some more details on what the mood changes do. The speed increase is something for example that should be documented in the PR description.

**Comment by Xackii** [comment]
> Please add some more details on what the mood changes do. The speed increase is something for example that should be documented in the PR description.

wrote it

**Comment by Xackii** [comment]
Ok, I accidentally deleted my fork of Paradise when I forked Paradise 220, so if anyone wants, you can re-upload the mood PR

**Pull request closed without merging**